docs/thruk.3
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "thruk 3"
.TH thruk 3 "2015-02-12" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
thruk \- Command Line Utility for Thruk Monitoring Gui
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  Usage: thruk [options] [url|files...]
\&
\&  Options:
\&  \-h, \-\-help                    Show this help message and exit
\&  \-v, \-\-verbose                 Print verbose output
\&  \-V, \-\-version                 Print version
\&
\&  \-b, \-\-backend=<backend>       Comma seperated list of backends/sites
\&                                which should be used for querys.
\&  \-l, \-\-list\-backends           List available backends/sites.
\&  \-a, \-\-action=<action>         Perform given action. see below for a list.
\&
\&  \-A, \-\-auth=<auth>             Set username for querys
\&
\&  \-c, \-\-credential=<pw>         Credential when connecting to remote
\&                                Thruk installations specified by the
\&                                remote\-url.
\&  \-r, \-\-remote\-url=<url>        URL to remote installation.
\&      \-\-local                   Don\*(Aqt try to connect fastcgi daemon first.
\&
\&  \-y, \-\-yes                     answer yes to all given questions.
\&  \-f, \-\-force                   some actions need to be forced
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This script gives access to various parts of the Thruk gui and allows to generate
excel reports, csv exports or any other url by using command line and bypassing the
webserver.
.SH "OPTIONS"
.IX Header "OPTIONS"
script has the following arguments
.IP "\fB\-h\fR , \fB\-\-help\fR" 4
.IX Item "-h , --help"
.Vb 1
\&    print help and exit
.Ve
.IP "\fB\-v\fR , \fB\-\-verbose\fR" 4
.IX Item "-v , --verbose"
.Vb 1
\&    print verbose output too
.Ve
.IP "\fB\-V\fR , \fB\-\-version\fR" 4
.IX Item "-V , --version"
.Vb 1
\&    print version and exit
.Ve
.IP "\fB\-b\fR \fIbackend\fR, \fB\-\-backend\fR=\fIbackend\fR" 4
.IX Item "-b backend, --backend=backend"
.Vb 3
\&    specify backend(s), otherwise the default backends will be
\&    used. Specify the key of the backend. Use <\-l> to get a list
\&    of backends.
\&
\&    All default backends (backends without the \*(Aqhidden\*(Aq flag) will be used unless
\&    backend(s) are specified with this option.
\&
\&    Use \*(AqALL\*(Aq to use all backends, regardless of the hidden flag.
.Ve
.IP "\fB\-l\fR , \fB\-\-list\-backends\fR" 4
.IX Item "-l , --list-backends"
.Vb 1
\&    get list of available backends
.Ve
.IP "\fB\-a\fR \fIaction\fR, \fB\-\-action\fR=\fIaction\fR" 4
.IX Item "-a action, --action=action"
.Vb 5
\&    perform selected action:
\&      \- url=<url>                   output url result
\&        \-i / \-\-all\-inclusive        includes all css, javascript and images in the resulting html page
\&      \- report=<nr>                 generate report
\&      \- reportmail=<nr>             send report by mail
\&
\&    Logcache: ( MySQL / MongoDB )
\&      \- logcacheimport(=blocksize)  initial logcache import. Blocksize is in seconds. Defaults is one day. (hint: use \-\-start=3w to import last 3weeks only)
\&      \- logcacheupdate              update logcache
\&      \- logcacheauthupdate          update logcache authorization information
\&      \- logcachestats               print logcache statistics
\&      \- logcacheclean(=days)        clean everything older than specified number of days. Default is one year.
\&      \- logcacheoptimize            optimize and repair logcache (mysql only)
\&      \- logcacheremoveunused        remove old tables from no longer used backends (mysql only)
\&
\&    Livestatus Cache:
\&      \- livecachestart              start shadownaemons livecache
\&      \- livecachestop               stop all shadownaemons
\&      \- livecacherestart            restart livecaches for each site sequentially
\&      \- livecachestatus             show livecache status
\&
\&    Internal Commands
\&      \- installcron                 update cronjobs for downtimes and reports
\&                                    (only needed during package installation)
\&      \- uninstallcron               remove all thruk cronjobs
\&                                    (only needed during package deinstallation)
\&      \- compile                     compile templates for faster access
\&      \- bp/bpd(=nr)                 calculate all/specified business process(es)
\&      \- bpcommit                    write objects for business processes and reloads core
\&
\&    Testing Commands
\&      \- listbackends                list all backends
\&      \- listhosts                   list hostnames
\&      \- listhostgroups              list hostgroupnames
\&      \- command <host> <service>    print used command line for this host/service
\&
\&    Selfcheck Commands
\&      \- selfcheck(=subtype)         run self check(s)
\&                                    availabel subtypes:
\&                                     \- filesystem
\&                                     \- logfiles
\&                                     \- reports
\&                                     \- recurring_downtimes
.Ve
.IP "\fB\-A\fR \fIauthname\fR, \fB\-\-auth\fR=\fIauthname\fR" 4
.IX Item "-A authname, --auth=authname"
.Vb 1
\&    use this username for authentication
.Ve
.IP "\fB\-c\fR \fIcredential\fR, \fB\-\-credential\fR=\fIcredential\fR" 4
.IX Item "-c credential, --credential=credential"
.Vb 1
\&    use this key for authentication when using remoteurl
.Ve
.IP "\fB\-r\fR \fIremote-url\fR, \fB\-\-remote\-url\fR=\fIremote-url\fR" 4
.IX Item "-r remote-url, --remote-url=remote-url"
.Vb 4
\&    Connect to this remote instance instead of the local one.
\&    Defaults to http://localhost/thruk/cgi\-bin/remote.cgi.
\&    Retrieving information from the fcgi daemon saves time to
\&    startup and allows to connect to remote located sites.
.Ve

Changes
This file documents the revision history for the Monitoring Webinterface Thruk.

next:
          - add host and service notes url macros (Andrew Widdersheim)
          - Bug Fixes
            - fix accessing performance data page in author mode
            - fix accessing reports page in author mode

1.86-4   Thu Feb 12 21:47:53 CET 2015:
          - add action_menu_apply option to apply action menus on the fly
          - Panorama View:
            - add group based permissions for dashboards
            - move multiple icons by selecting them with ctrl click
            - also move multiple icons by mouse lasso
          - Bug Fixes
            - fix using comma separated backend list in CLI tool
            - fix undefined key_sort in CLI config parser

1.86-3   Thu Jan 29 19:57:35 CET 2015
          - add option for default service filter
          - add host/servce connection macros
          - Bug Fixes
            - fix icon logo path in statusmap parent view
            - fixed links to other dashboards in panorama plugin
            - fix listing all backends with object config when using http backends
            - fix product prefix in cli url requests

1.86-2   Sun Jan 18 17:45:37 CET 2015
          - Reporting:
            - add option for maximum number of concurrent reports
            - start reports from cron serial instead of all at once
          - Panorama View:
            - add server actions as possible link targets
            - fixed problem when moving rotated shapes
          - Bug Fixes
            - fixed macro replacement in action menu links
            - fixed error in cluetip library by shiping a new version
            - fixed problem with shadownaemon in non-xs env
            - fixed error: Can't use string xxx as an ARRAY

1.86     Thu Jan  8 14:10:41 CET 2015
          - add custom action menu
          - livestatus performance improvements
          - updated jquery to 1.11.1
          - added new configuration setting: max_process_memory
          - add excel export permanent link (awiddersheim)
          - increased logcache int size
          - use grey performance bars on extinfo page
          - Business Process:
            - fix zoom on large business objects
          - Mobile:
            - updated jquery mobile
            - respect escape_html_tags setting from cgi.cfg
            - old themes have been replaced with 2 new ones
          - Reporting:
            - support url reports from external urls
          - Panorama View:
            - added show details/refresh links to icons
            - add 'center' position option for labels
            - changeable icon filter types
            - fixed memory leak in generic url panel
            - fixed problem with stopped timers after import
            - reload every x hours to prevent noncatchable memory leaks
            - fixed linking public dashboards
          - Bug Fixes
            - fixed problem with timeperiods in logcache database
            - fixed show_custom_vars with wildcards
            - fixed adding dummy function in business processes
            - fixed startup page in safari browser
            - fixed css problem in mobile client

1.84-6   Wed Aug  6 11:24:12 CEST 2014
          - Panorama View:
            - support availability results in labels and speedometer

1.84-5   Wed Jul 16 13:30:05 CEST 2014
          - Reports
            - added cancel button for running reports
            - added progres bar for running reports
          - Panorama View:
            - added -or filter to all filter panels
            - added optional min/max values for speedometer
            - added invert switch for speedometer
            - background images can now be scaled
            - background images can have an offset
            - images and icons can be scaled
          - Bug Fixes
            - fixed pnp source not selectable anymore after revisiting panel settings
            - fixed applying panel source
            - fixed using multiple panorama custom filter

1.84-4   Mon Jun 23 17:44:32 CEST 2014
          - added show_error_reports mode 'both' which shows the error indicator but logs to server side
          - set default show_error_reports mode to 'both' to make error handling easier
          - support pnp urls with suffix
          - Panorama View:
            - added z-index option for icons
          - Bug Fixes
            - fixed rendering icons above dashboard settings window
            - fixed recurring downtimes not beeing saved under some conditions with multiple backends
            - fixed problem in servicegroup recurring downtimes
            - fixed half-visible panorama buttons in IE11
            - fixed panorama shape rotation in IE11
            - fixed panorama column error in minemap
            - fixed problem with utf8 characters in passwords

lib/Thruk.pm
package Thruk;

=head1 NAME

Thruk - Catalyst based monitoring web interface

=head1 DESCRIPTION

Catalyst based monitoring web interface for Nagios, Icinga and Shinken

=cut

use 5.008000;
use strict;
use warnings;
use utf8;

###################################################
# create connection pool
# has to be done before the binmode
# or even earlier to save memory
use Thruk::Backend::Pool;
BEGIN {
    Thruk::Backend::Pool::init_backend_thread_pool();
};

###################################################
# load timing class
BEGIN {
    #use Thruk::Timer qw/timing_breakpoint/;
    #&timing_breakpoint('starting thruk');
};

###################################################
# clean up env
use Thruk::Utils::INC;
BEGIN {
    Thruk::Utils::INC::clean();
    ## no critic
    eval "use Time::HiRes qw/gettimeofday tv_interval/;" if ($ENV{'THRUK_PERFORMANCE_DEBUG'} and $ENV{'THRUK_PERFORMANCE_DEBUG'} >= 0);
    eval "use Thruk::Template::Context;"                 if ($ENV{'THRUK_PERFORMANCE_DEBUG'} and $ENV{'THRUK_PERFORMANCE_DEBUG'} >= 3);
    ## use critic
    #&timing_breakpoint('cleaned env');
}
use Carp;
use Moose;
use GD;
use POSIX qw(tzset);
use Log::Log4perl::Catalyst;
use Digest::MD5 qw(md5_hex);
use File::Slurp qw(read_file);
use Data::Dumper;
use MRO::Compat;
use Thruk::Utils;
use Thruk::Config;
use Thruk::Utils::Auth;
use Thruk::Utils::Filter;
use Thruk::Utils::Menu;
use Thruk::Utils::Avail;
use Thruk::Utils::External;
use Thruk::Utils::Livecache;
use Thruk::Utils::Cache qw/cache/;
use Catalyst::Runtime '5.70';

###################################################
# Set flags and add plugins for the application
#
#         -Debug: activates the debug mode for very useful log messages
#         StackTrace
BEGIN {
    #&timing_breakpoint('begin thruk');
    # Compress - temporarily disabled till https://rt.cpan.org/Ticket/Display.html?id=87998 gets fixed
    my @catalyst_plugins = qw/
          Thruk::ConfigLoader
          Unicode::Encoding
          Authentication
          Authorization::ThrukRoles
          CustomErrorMessage
          Static::Simple
          Redirect
          Thruk::RemoveNastyCharsFromHttpParam
    /;
    if($Catalyst::Runtime::VERSION >= 5.90042) {
        # since 5.90042 catalyst encodes by core
        @catalyst_plugins = grep(!/^Unicode::Encoding$/mx, @catalyst_plugins);
    }
    # fcgid setups have no static content
    if(defined $ENV{'THRUK_SRC'} and $ENV{'THRUK_SRC'} eq 'FastCGI') {
        @catalyst_plugins = grep(!/^Static::Simple$/mx, @catalyst_plugins);
    }
    #&timing_breakpoint('loading Catalyst');
    require Catalyst;
    Catalyst->import(@catalyst_plugins);
    __PACKAGE__->config( encoding => 'UTF-8' );
    #&timing_breakpoint('loading Catalyst done');
};

###################################################
our $VERSION = '1.86';

###################################################
# load config loader
__PACKAGE__->config(%Thruk::Config::config);
__PACKAGE__->config( encoding => 'UTF-8' );
#&timing_breakpoint('config loaded');

###################################################
# install leak checker
if($ENV{THRUK_LEAK_CHECK}) {
    eval {
        with 'CatalystX::LeakChecker';
        $Devel::Cycle::already_warned{'GLOB'} = 1;
    };
    print STDERR "failed to load CatalystX::LeakChecker: ".$@ if $@;
}

###################################################
# Start the application and make __PACKAGE__->config
# accessible
# override config in Catalyst::Plugin::Thruk::ConfigLoader
__PACKAGE__->setup();
$Thruk::Utils::IO::config = __PACKAGE__->config;
#&timing_breakpoint('setup done');

###################################################
binmode(STDOUT, ":encoding(UTF-8)");
binmode(STDERR, ":encoding(UTF-8)");
$Data::Dumper::Sortkeys = 1;

###################################################
# init cache
Thruk::Utils::IO::mkdir(__PACKAGE__->config->{'tmp_path'});
__PACKAGE__->cache(__PACKAGE__->config->{'tmp_path'}.'/thruk.cache');

###################################################
# save pid
my $pidfile  = __PACKAGE__->config->{'tmp_path'}.'/thruk.pid';
sub _remove_pid {
    $SIG{PIPE} = 'IGNORE';
    if(defined $ENV{'THRUK_SRC'} and $ENV{'THRUK_SRC'} eq 'FastCGI') {
        if($pidfile && -f $pidfile) {
            my $pids = [split(/\s/mx, read_file($pidfile))];
            my $remaining = [];
            for my $pid (@{$pids}) {
                next unless($pid and $pid =~ m/^\d+$/mx);
                next if $pid == $$;
                next if kill(0, $pid) == 0;
                push @{$remaining}, $pid;
            }
            if(scalar @{$remaining} == 0) {
                unlink($pidfile);
                if(__PACKAGE__->config->{'use_shadow_naemon'} and __PACKAGE__->config->{'use_shadow_naemon'} ne 'start_only') {
                    Thruk::Utils::Livecache::shutdown_shadow_naemon_procs(__PACKAGE__->config);
                }
            } else {
                open(my $fh, '>', $pidfile);
                print $fh join("\n", @{$remaining}),"\n";
                CORE::close($fh);
            }
        }
    }
    if(defined $ENV{'THRUK_SRC'} and $ENV{'THRUK_SRC'} eq 'DebugServer') {
        # debug server has no pid file, so just kill our shadows
        if(__PACKAGE__->config->{'use_shadow_naemon'} and __PACKAGE__->config->{'use_shadow_naemon'} ne 'start_only') {
            Thruk::Utils::Livecache::shutdown_shadow_naemon_procs(__PACKAGE__->config);
        }
    }
    return;
}
if(defined $ENV{'THRUK_SRC'} and $ENV{'THRUK_SRC'} eq 'FastCGI') {
    -s $pidfile || unlink(__PACKAGE__->config->{'tmp_path'}.'/thruk.cache');
    open(my $fh, '>>', $pidfile) || warn("cannot write $pidfile: $!");
    print $fh $$."\n";
    Thruk::Utils::IO::close($fh, $pidfile);
}
$SIG{INT}  = sub { _remove_pid(); exit; };
$SIG{TERM} = sub { _remove_pid(); exit; };
END {
    _remove_pid();
};

###################################################
# create secret file
if(!defined $ENV{'THRUK_SRC'} or $ENV{'THRUK_SRC'} ne 'SCRIPTS') {
    my $var_path   = __PACKAGE__->config->{'var_path'} or die("no var path!");
    my $secretfile = $var_path.'/secret.key';
    unless(-s $secretfile) {
        my $digest = md5_hex(rand(1000).time());
        chomp($digest);
        open(my $fh, ">$secretfile") or warn("cannot write to $secretfile: $!");
        if(defined $fh) {
            print $fh $digest;
            Thruk::Utils::IO::close($fh, $secretfile);
            chmod(0640, $secretfile);
        }
        __PACKAGE__->config->{'secret_key'} = $digest;
    } else {
        my $secret_key = read_file($secretfile);
        chomp($secret_key);
        __PACKAGE__->config->{'secret_key'} = $secret_key;
    }
}

###################################################
# set timezone
my $timezone = __PACKAGE__->config->{'use_timezone'};
if(defined $timezone) {
    $ENV{'TZ'} = $timezone;
    POSIX::tzset();
}

###################################################
# set installed server side includes
my $ssi_dir = __PACKAGE__->config->{'ssi_path'};
my (%ssi, $dh);
if(!-e $ssi_dir) {
    warn("cannot access ssi_path $ssi_dir: $!");
} else {
    opendir( $dh, $ssi_dir) or die "can't opendir '$ssi_dir': $!";
    for my $entry (readdir($dh)) {
        next if $entry eq '.' or $entry eq '..';
        next if $entry !~ /\.ssi$/mx;
        $ssi{$entry} = { name => $entry }
    }
    closedir $dh;
}
__PACKAGE__->config->{'ssi_includes'} = \%ssi;
__PACKAGE__->config->{'ssi_path'}     = $ssi_dir;

###################################################
# load and parse cgi.cfg into $c->config
unless(Thruk::Utils::read_cgi_cfg(undef, __PACKAGE__->config)) {
    die("\n\n*****\nfailed to load cgi config: ".__PACKAGE__->config->{'cgi.cfg'}."\n*****\n\n");
}


###################################################
# Logging
my $log4perl_conf;
if(!defined $ENV{'THRUK_SRC'} or ($ENV{'THRUK_SRC'} ne 'CLI' and $ENV{'THRUK_SRC'} ne 'SCRIPTS')) {
    if(defined __PACKAGE__->config->{'log4perl_conf'} and ! -s __PACKAGE__->config->{'log4perl_conf'} ) {
        die("\n\n*****\nfailed to load log4perl config: ".__PACKAGE__->config->{'log4perl_conf'}.": ".$!."\n*****\n\n");
    }
    $log4perl_conf = __PACKAGE__->config->{'log4perl_conf'} || __PACKAGE__->config->{'home'}.'/log4perl.conf';
}
if(defined $log4perl_conf and -s $log4perl_conf) {
    __PACKAGE__->log(Log::Log4perl::Catalyst->new($log4perl_conf));
    __PACKAGE__->config->{'log4perl_conf_in_use'} = $log4perl_conf;
}
elsif(!__PACKAGE__->debug) {
    __PACKAGE__->log->levels( 'info', 'warn', 'error', 'fatal' );
}

###################################################
# SizeMe and other devel internals
if($ENV{'SIZEME'}) {
    # add signal handler to print memory information
    # ps -efl | grep perl | grep thruk_server.pl | awk '{print $4}' | xargs kill -USR1
    $SIG{'USR1'} = sub {
        printf(STDERR "mem:% 7s MB  before devel::sizeme\n", Thruk::Backend::Pool::get_memory_usage());
        eval {
            require Devel::SizeMe;
            Devel::SizeMe::perl_size();
        };
        print STDERR $@ if $@;
    }
}
if($ENV{'MALLINFO'}) {
    # add signal handler to print memory information
    # ps -efl | grep perl | grep thruk_server.pl | awk '{print $4}' | xargs kill -USR2
    $SIG{'USR2'} = sub {
        eval {
            require Devel::Mallinfo;
            require Data::Dumper;
            my $info = Devel::Mallinfo::mallinfo();
            printf STDERR "%s\n", '*******************************************';
            printf STDERR "%-30s    %5.1f %2s\n", 'arena',                              Thruk::Utils::reduce_number($info->{'arena'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'bytes in use, ordinary blocks',  Thruk::Utils::reduce_number($info->{'uordblks'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'bytes in use, small blocks',     Thruk::Utils::reduce_number($info->{'usmblks'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'free bytes, ordinary blocks',    Thruk::Utils::reduce_number($info->{'fordblks'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'free bytes, small blocks',       Thruk::Utils::reduce_number($info->{'fsmblks'}, 'B');
            printf STDERR "%-30s\n", 'total';
            printf STDERR "   %-30s %5.1f %2s\n", 'taken from the system',    Thruk::Utils::reduce_number($info->{'arena'} + $info->{'hblkhd'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'in use by program',        Thruk::Utils::reduce_number($info->{'uordblks'} + $info->{'usmblks'} + $info->{'hblkhd'}, 'B');
            printf STDERR "   %-30s %5.1f %2s\n", 'free within program',      Thruk::Utils::reduce_number($info->{'fordblks'} + $info->{'fsmblks'}, 'B');
        };
        print STDERR $@ if $@;
    }
}

###################################################

=head1 METHODS

=head2 check_user_roles_wrapper

  check_user_roles_wrapper()

wrapper to avoid undef values in TT

=cut
sub check_user_roles_wrapper {
    my $self = shift;
    if($self->check_user_roles(@_)) {
        return 1;
    }
    return 0;
}

###################################################

=head2 found_leaks

called by CatalystX::LeakChecker and used for testing purposes only

=cut
sub found_leaks {
    my ($c, @leaks) = @_;
    return unless scalar @leaks > 0;
    my $sym = 'a';
    print STDERR "found leaks:\n";
    for my $leak (@leaks) {
        my $msg = (CatalystX::LeakChecker::format_leak($leak, \$sym));
        $c->log->error($msg);
        print STDERR $msg,"\n";
    }
    if(defined $ENV{'THRUK_SRC'} and $ENV{'THRUK_SRC'} eq 'TEST_LEAK') {
        die("tests die, exit otherwise");
    }
    # die() won't let our tests exit, so we use exit here
    exit 1;
    return;
}

###################################################

=head2 prepare_path

called by catalyst to strip path prefixes

=cut
sub prepare_path {
    my($c) = @_;
    $c->maybe::next::method();

    #&timing_breakpoint('request '.$c->request->path) if $c->request->path =~ m/cgi\-bin|html/;

    # collect statistics when running external command or if enabled by env variable
    if($ENV{'THRUK_JOB_DIR'} || ($ENV{'THRUK_PERFORMANCE_DEBUG'} && $ENV{'THRUK_PERFORMANCE_DEBUG'} >= 2)) {
        $c->stats->enable(1);
    }

    my $product = $c->config->{'product_prefix'} || 'thruk';
    if($product ne 'thruk') {
        # make it look like a thruk url, ex.: .../thruk/cgi-bin/tac.cgi
        my $path = $c->request->path;
        my $product_prefix = $c->config->{'product_prefix'};
        $path =~ s|^\Q$product_prefix\E||mxo;
        $path = $path ? "thruk/".$path : 'thruk';
        $path =~ s|thruk/+|thruk/|gmxo;
        $c->request->path($path);
        return;
    }

    # for OMD
    my @path_chunks = split m[/]mxo, $c->request->path, -1;
    return unless($path_chunks[1] and $path_chunks[1] eq 'thruk');

    my $site = shift @path_chunks;
    my $path = join('/', @path_chunks) || '/';
    $c->request->path($path);
    my $base = $c->request->base;
    $base->path($base->path.$site.'/');

    return;
}

###################################################

=head2 run_after_request

run callbacks after the request had been send to the client

=cut
sub run_after_request {
    my ($c, $sub) = @_;
    $c->stash->{'run_after_request_cb'} = [] unless defined $c->stash->{'run_after_request_cb'};
    push @{$c->stash->{'run_after_request_cb'}}, $sub;
    return;
}

before finalize => sub {
    my($c) = @_;
    $c->stash->{'no_more_profile'} = 1;
    $c->stats->profile(begin => "finalize");
    return;
};

after finalize => sub {
    my($c) = @_;
    $c->stats->profile(end => "finalize");
    $c->stats->profile(begin => "after finalize");

    while(my $sub = shift @{$c->stash->{'run_after_request_cb'}}) {
        ## no critic
        eval($sub);
        ## use critic
        $c->log->info($@) if $@;
    }
    $c->stats->profile(end => "after finalize");

    if($ENV{'THRUK_PERFORMANCE_DEBUG'} and $c->stash->{'memory_begin'}) {
        my $elapsed = tv_interval($c->stash->{'time_begin'});
        $c->stash->{'memory_end'} = Thruk::Backend::Pool::get_memory_usage();
        my($url) = ($c->request->uri =~ m#.*?/thruk/(.*)#mxo);
        $url     = $c->request->uri unless $url;
        $url     =~ s/^cgi\-bin\///mxo;
        if(length($url) > 50) { $url = substr($url, 0, 50).'...' }
        $c->log->info(sprintf("Req: %03d, mem:% 7s MB  % 10.2f MB     %.2fs %8s    %s\n",
                                ($Catalyst::COUNT || 0),
                                $c->stash->{'memory_end'},
                                ($c->stash->{'memory_end'}-$c->stash->{'memory_begin'}),
                                $elapsed,
                                defined $c->stash->{'total_backend_waited'} ? sprintf('(%.2fs)', $c->stash->{'total_backend_waited'}) : '',
                                $url,
                    ));
    }

    if($ENV{'THRUK_PERFORMANCE_DEBUG'} and $ENV{'THRUK_PERFORMANCE_DEBUG'} >= 2) {
        Thruk::Utils::External::log_profile($c);
    }

    Thruk::Utils::External::save_profile($c, $ENV{'THRUK_JOB_DIR'}) if $ENV{'THRUK_JOB_DIR'};
    return;
};

###################################################

=head2 use_stats

switch for various internal catalyst sub wether to gather statistics or not

=cut
sub use_stats {
    my($c) = @_;
    eval { # newer Catalyst::Middleware::Stash versions die if called to early
        if($c->stash->{'no_more_profile'})  { return(0); }
    };
    if($ENV{'THRUK_PERFORMANCE_DEBUG'} and $ENV{'THRUK_PERFORMANCE_DEBUG'} > 1) { return(1); }
    return(0);
}

###################################################
# add some more profiles
before prepare_body     => sub {
    my($c) = @_;
    if($ENV{'THRUK_PERFORMANCE_DEBUG'}) {
        $c->stash->{'memory_begin'} = Thruk::Backend::Pool::get_memory_usage();
        $c->stash->{'time_begin'}   = [gettimeofday()];
    }
    $c->stats->profile(begin => "prepare_body");
    return;
};
after prepare_body      => sub { my($c) = @_; $c->stats->profile(end   => "prepare_body");     return; };

before dispatch         => sub { my($c) = @_; $c->stats->profile(begin => "dispatch");         return; };
after dispatch          => sub { my($c) = @_; $c->stats->profile(end   => "dispatch");         return; };

before finalize_uploads => sub { my($c) = @_; $c->stats->profile(begin => "finalize_uploads"); return; };
after finalize_uploads  => sub { my($c) = @_; $c->stats->profile(end   => "finalize_uploads"); return; };

before finalize_error   => sub { my($c) = @_; $c->stats->profile(begin => "finalize_error");   return; };
after finalize_error    => sub { my($c) = @_; $c->stats->profile(end   => "finalize_error");   return; };

before finalize_headers => sub { my($c) = @_; $c->stats->profile(begin => "finalize_headers"); return; };
after finalize_headers  => sub { my($c) = @_; $c->stats->profile(end   => "finalize_headers"); return; };

before finalize_cookies => sub { my($c) = @_; $c->stats->profile(begin => "finalize_cookies"); return; };
after finalize_cookies  => sub { my($c) = @_; $c->stats->profile(end   => "finalize_cookies"); return; };

before finalize_body    => sub { my($c) = @_; $c->stats->profile(begin => "finalize_body");    return; };
after finalize_body     => sub { my($c) = @_; $c->stats->profile(end   => "finalize_body");    return; };

#&timing_breakpoint('start done');

=head1 SEE ALSO

L<Thruk::Controller::Root>, L<Catalyst>

=head1 AUTHOR

Sven Nierlein, 2009-2014, <sven@nierlein.org>

=head1 LICENSE

This library is free software, you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;
